libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2476,11 +2476,8 @@ string YulUtilFunctions::bytesConcatFunction(vector<Type const*> const& _argumen
 			literalType && literalType->value().size() <= 32
 		)
 			targetTypes.emplace_back(TypeProvider::fixedBytes(static_cast<unsigned>(literalType->value().size())));
-		else if (auto const* literalType = dynamic_cast<RationalNumberType const*>(argumentType))
-		{
-			solAssert(literalType->value() == 0, "");
-			targetTypes.emplace_back(TypeProvider::fixedBytes(1));
-		}
+		else if (argumentType->isImplicitlyConvertibleTo(*TypeProvider::fixedBytes(32)))
+			targetTypes.emplace_back(TypeProvider::fixedBytes(32));
```

Comment:
Wait are we always using `bytes32`? Was it like that before? Is it consistent with `abi.encodePacked()`?

Reply:
Only for literals. Well, I think I should add a check that it's actually `RationalNumberType` here to make it 100% clear but currently that's the only possibility anyway. At this point all fixed bytes types and string literals have already been handled by `if` branches above and `bytes` is not implicitly convertible to fixed `bytes32` so it will skip this one.

Reply:
I added a stronger check now with an assertion that shows that it only happens for literals that are compatible with `bytes32`.

> Was it like that before?

Literals were not considered here at all before. They were falling into the `else` branch that handles `bytes` and failing the assertion there.

> Is it consistent with `abi.encodePacked()`?

It is unless `abi.encodePacked()` handles 32-byte hex literals as something else than `bytes32`. Not sure what it does with zero. I'll check that.

Reply:
I checked `abi.encodePacked()` and it does not allow 32-byte hex literals at all:
```
Error: Cannot perform packed encoding for a literal. Please convert it to an explicit type first.
  --> test.sol:53:13:
   |
53 |             0x00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Does not allow plain zero either. Only allows various flavors of string literals (`hex"aabbcc"`, `unicode"abc"`, `"abc"`).

But it also allows some of the literals not supported by `bytes.concat()`:
- Boolean literals (`false`, `true`).
- Address literals (`0x1111222233334444555566667777888899990000`).
- Array literals (`[0]`).
- Expressions that yield integers (`[0][0]`).
- Enum members (`E.A`).

Reply:
Then let's just disallow literal numbers. I think it's fair to require explicit conversion to a bytes type for all the examples above.

Reply:
Done.